### ISSUE TITLE
XLA TPU, argmax return wrong value when softmax is applied

### ISSUE THREAD

**Issue by Shiro-LK** [pytorch/xla#2525]
## 🐛 Bug

When using torch.argmax or torch.max to get the index of the max value of a tensor, the value return are constant if softmax was applied before the argmax.
## To Reproduce

https://colab.research.google.com/drive/1F1Jez9raGSJ0tRcaKsCCK-LqfPvgz6ws#scrollTo=_2nL4HmloEyl

1. activate bfloat16 : `os.environ['XLA_USE_BF16']="1"`
2. use softmax `T = torch.softmax(T, dim=-1)`
3. use argmax :` torch.max(T, dim=-1)[1]
`

## Expected behavior
exemple of wrong value : 
> tensor([[-9223372036854775808, -9223372036854775808, -9223372036854775808,
>           ..., -9223372036854775808, -9223372036854775808,
>          -9223372036854775808],
>         [-9223372036854775808, -9223372036854775808, -9223372036854775808,
>           ..., -9223372036854775808, -9223372036854775808,
>          -9223372036854775808],
>         [-9223372036854775808, -9223372036854775808, -9223372036854775808,
>           ..., -9223372036854775808, -9223372036854775808,
>          -9223372036854775808],
>         ...,
>         [-9223372036854775808, -9223372036854775808, -9223372036854775808,
>           ..., -9223372036854775808, -9223372036854775808,
>          -9223372036854775808],
>         [-9223372036854775808, -9223372036854775808, -9223372036854775808,
>           ..., -9223372036854775808, -9223372036854775808,
>          -9223372036854775808],
>         [-9223372036854775808, -9223372036854775808, -9223372036854775808,
>           ..., -9223372036854775808, -9223372036854775808,
>          -9223372036854775808]], device='xla:1')

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU only
 - torch_xla version: 1.6 but the issue is also present in nightly version

**Comment by JackCaoG** [comment]
Thanks for reporting @Shiro-LK ! I am able to repo your problem on TPU and it seems to be a bug on the xla side. 
```
tensor([[-7.4423,  8.3796, -5.9036, -7.1343,  5.7470],
        [-6.9329, -4.7965, -1.9988, -3.0048, -7.2666],
        [ 7.8274,  8.7062,  8.4518,  3.6965, -0.0157],
        [ 0.3176,  8.1242,  9.1028,  5.9165,  0.8806],
        [-4.6448, -7.5289, -6.8786,  7.9011,  7.6976]], device='xla:1',
       dtype=torch.bfloat16)
tensor([[1.2683e-07, 9.3431e-01, 5.8644e-07, 1.7335e-07, 6.7681e-02],
        [4.9626e-03, 4.2870e-02, 6.9189e-01, 2.5453e-01, 3.5190e-03],
        [1.8921e-01, 4.5390e-01, 3.5350e-01, 3.1065e-03, 7.5377e-05],
        [1.0648e-04, 2.6160e-01, 7.1111e-01, 2.8448e-02, 1.8651e-04],
        [1.9399e-06, 1.0944e-07, 2.1095e-07, 5.5411e-01, 4.4524e-01]],
       device='xla:1', dtype=torch.bfloat16)
tensor([0.9336, 0.6914, 0.4531, 0.7109, 0.5547], device='xla:1',
       dtype=torch.bfloat16)
tensor([9223372036854775807, 9223372036854775807, 9223372036854775807,
        9223372036854775807, 9223372036854775807], device='xla:1')
```
If I turn of the optimization(fusion of ops) by 
```
$ export XLA_SYNC_TENSORS_OPBYOP=1
$ export XLA_GET_TENSORS_OPBYOP=1
```
I am seeing the correct result
```
tensor([[ 2.5469, -2.4062,  5.2188, -6.9062,  0.3809],
        [-5.1562,  4.3438,  1.8516, -1.7266, -1.5078],
        [ 3.0156, -4.2500,  6.8438,  8.3125,  4.5625],
        [-6.8438, -7.8750,  9.5625,  3.2656, -2.6250],
        [-6.0938,  2.5938, -1.5859, -7.5312, -6.8750]], device='xla:1',
       dtype=torch.bfloat16)
tensor([[6.3965e-02, 4.5204e-04, 9.2578e-01, 5.0366e-06, 7.3547e-03],
        [6.9141e-05, 9.2188e-01, 7.6172e-02, 2.1210e-03, 2.6550e-03],
        [3.9673e-03, 2.7865e-06, 1.8262e-01, 7.9688e-01, 1.8677e-02],
        [7.4971e-08, 2.6776e-08, 1.0000e+00, 1.8387e-03, 5.0962e-06],
        [1.6594e-04, 9.8438e-01, 1.5076e-02, 3.9339e-05, 7.5817e-05]],
       device='xla:1', dtype=torch.bfloat16)
tensor([0.9258, 0.9219, 0.7969, 1.0000, 0.9844], device='xla:1',
       dtype=torch.bfloat16)
tensor([2, 1, 3, 2, 1], device='xla:1')
```

9223372036854775807 is int64_max, but I can't guess what the bug might be.. I will open a bug on the xla side and update this thread when I hear back.

In the meantime, I can see couple workarounds
1. don't use the `XLA_USE_BF16 ` flag (recommended)
2. Use the OPBYOP mode (will damage the performance by quit a lot)

**Comment by JackCaoG** [comment]
I dumped the HLO and run it by itself and see the correct result being printed out.. so this is not a xla problem. Needs to debug a bit further.
```
HloModule IrToHlo.44

%MaxComputation.3 (x.4: bf16[], y.5: bf16[]) -> bf16[] {
  %x.4 = bf16[] parameter(0)
  %y.5 = bf16[] parameter(1)
  ROOT %maximum.6 = bf16[] maximum(bf16[] %x.4, bf16[] %y.5)
}

%AddComputation.12 (x.13: bf16[], y.14: bf16[]) -> bf16[] {
  %x.13 = bf16[] parameter(0)
  %y.14 = bf16[] parameter(1)
  ROOT %add.15 = bf16[] add(bf16[] %x.13, bf16[] %y.14)
}

%MaxComputation.21 (x.22: bf16[], y.23: bf16[]) -> bf16[] {
  %x.22 = bf16[] parameter(0)
  %y.23 = bf16[] parameter(1)
  ROOT %maximum.24 = bf16[] maximum(bf16[] %x.22, bf16[] %y.23)
}

%max_BF16.28 (lhs.29: bf16[], rhs.30: bf16[]) -> bf16[] {
  %lhs.29 = bf16[] parameter(0)
  %rhs.30 = bf16[] parameter(1)
  ROOT %maximum.31 = bf16[] maximum(bf16[] %lhs.29, bf16[] %rhs.30)
}

%min_S64.38 (lhs.39: s64[], rhs.40: s64[]) -> s64[] {
  %lhs.39 = s64[] parameter(0)
  %rhs.40 = s64[] parameter(1)
  ROOT %minimum.41 = s64[] minimum(s64[] %lhs.39, s64[] %rhs.40)
}

ENTRY %IrToHlo.44 (p0.1: bf16[5,5]) -> (s64[5]) {
  %constant.20 = s32[] constant(5), metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %p0.1 = bf16[5,5]{1,0} parameter(0), metadata={op_type="xla::device_data" source_file="<module>@test.py" source_line=11}
  %constant.2 = bf16[] constant(-inf), metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %reduce.7 = bf16[5]{0} reduce(bf16[5,5]{1,0} %p0.1, bf16[] %constant.2), dimensions={1}, to_apply=%MaxComputation.3, metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %broadcast.8 = bf16[5,5]{1,0} broadcast(bf16[5]{0} %reduce.7), dimensions={0}, metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %subtract.9 = bf16[5,5]{1,0} subtract(bf16[5,5]{1,0} %p0.1, bf16[5,5]{1,0} %broadcast.8), metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %exponential.10 = bf16[5,5]{1,0} exponential(bf16[5,5]{1,0} %subtract.9), metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %constant.11 = bf16[] constant(0), metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %reduce.16 = bf16[5]{0} reduce(bf16[5,5]{1,0} %exponential.10, bf16[] %constant.11), dimensions={1}, to_apply=%AddComputation.12, metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %broadcast.17 = bf16[5,5]{1,0} broadcast(bf16[5]{0} %reduce.16), dimensions={0}, metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %divide.18 = bf16[5,5]{1,0} divide(bf16[5,5]{1,0} %exponential.10, bf16[5,5]{1,0} %broadcast.17), metadata={op_type="aten::softmax" source_file="<module>@test.py" source_line=11}
  %constant.19 = bf16[] constant(-inf), metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %reduce.25 = bf16[5]{0} reduce(bf16[5,5]{1,0} %divide.18, bf16[] %constant.19), dimensions={1}, to_apply=%MaxComputation.21, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %constant.26 = bf16[] constant(-inf), metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %reduce.32 = bf16[5]{0} reduce(bf16[5,5]{1,0} %divide.18, bf16[] %constant.26), dimensions={1}, to_apply=%max_BF16.28, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %broadcast.34 = bf16[5,5]{1,0} broadcast(bf16[5]{0} %reduce.32), dimensions={0}, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %compare.35 = pred[5,5]{1,0} compare(bf16[5,5]{1,0} %divide.18, bf16[5,5]{1,0} %broadcast.34), direction=EQ, type=UNSIGNED, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %iota.27 = s64[5,5]{1,0} iota(), iota_dimension=1, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %constant.33 = s64[] constant(9223372036854775807), metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %broadcast.36 = s64[5,5]{1,0} broadcast(s64[] %constant.33), dimensions={}, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %select.37 = s64[5,5]{1,0} select(pred[5,5]{1,0} %compare.35, s64[5,5]{1,0} %iota.27, s64[5,5]{1,0} %broadcast.36), metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  %reduce.42 = s64[5]{0} reduce(s64[5,5]{1,0} %select.37, s64[] %constant.33), dimensions={1}, to_apply=%min_S64.38, metadata={op_type="aten::max" source_file="<module>@test.py" source_line=13}
  ROOT %tuple.43 = (s64[5]{0}) tuple(s64[5]{0} %reduce.42)
}

```
```
** Result on test platform TPU_Interpreter **
(
s64[5] {0, 1, 0, 3, 4}
)
```

**Comment by JackCaoG** [comment]
For me, adding a `xm.mark_step()`(which will execute the pending graph and materialize all of the live tensor) before `print()` will solve this problem.  You can workaround it by this for now while I am investigating this bug. It is recommended to call `xm.mark_step()` before looking at the tensor value, more detail in [here](https://github.com/pytorch/xla/blob/9e9bc80ccf1eaa10668a811b8fdb71d98b67730a/API_GUIDE.md#L103).

**Comment by Shiro-LK** [comment]
Thank you I will try it 👍

**Comment by JackCaoG** [comment]
I used
```
import torch, torch_xla                                                                                                                                                                                                      
import os                                                                                                                                                                                                                    
import torch_xla.core.xla_model as xm                                                                                                                                                                                        
                                                                                                                       
device = xm.xla_device()                                                                                                                                                                                                     
                                                                                                                                                                                                                             
T = torch.FloatTensor(5,5).uniform_(-10, 10).to(torch.bfloat16).to(device)                                                                                                                                                   
T = torch.softmax(T, dim=-1)                                                                                                                                                                                                 
max_proba, idx = torch.max(T, dim=-1)                                                                                                                                                                                        
                                                                                                                                                                                                                       
                                                                                                                                                                                
print(idx.cpu())                                                                                                                                                                                                             
xm.mark_step()                                                                                                                                                                                                               
print(idx.cpu())
print(torch.max(T, dim=-1)[1].cpu())
```
to repo the issue and I saw the output

```
tensor([9223372036854775807, 9223372036854775807, 9223372036854775807,
        9223372036854775807, 9223372036854775807])
tensor([9223372036854775807, 9223372036854775807, 9223372036854775807,
        9223372036854775807, 9223372036854775807])
tensor([4, 1, 1, 3, 1])
```

It seems like after `mark_step`, `idx` value would not be changed but rerun the compuation will give the correct result. 

ir graph of `idx` also seems correct
```
(Pdb) print(torch_xla._XLAC._get_xla_tensors_text([idx]))
IR {
  %0 = bf16[5,5]{1,0} xla::device_data(), location=<module>@test.py:10, device=TPU:0
  %1 = bf16[] prim::Constant(), location=<module>@test.py:10, value=0
  %2 = bf16[5,5]{1,0} aten::expand(%1), location=<module>@test.py:10, size=(5, 5)
  %3 = bf16[5,5]{1,0} xla::as_strided_view_update(%2, %0), location=<module>@test.py:11, size=(5, 5), stride=(5, 1), storage_offset=0
  %4 = bf16[5,5]{1,0} aten::as_strided(%3), location=<module>@test.py:11, size=(5, 5), stride=(5, 1), storage_offset=0
  %5 = bf16[5,5]{1,0} aten::softmax(%4), location=<module>@test.py:11, dim=1, dtype=15
  %6 = (bf16[5]{0}, s64[5]{0}) aten::max(%5), num_outputs=2, location=<module>@test.py:12, dim=1, keepdim=0, ROOT=0
}
```

I follow the gdb and found that `print` will trigger the `toTensor` and `applyPendingGraph` for the `idx` tensor. For some reason this exeuction will give the incorrect result (all `9223372036854775807`). `mark_step` will call `SyncLiveTensors` and sync all tensors, it will give the correct result.

This corner case also has something to do with the size of the input tensor, if I changed it to (4,5) instead of (5,5), this bug will not happen..

**Comment by JackCaoG** [comment]
This actually have to do with precision, so we only see it on TPU with bfloat 16

PT/XLA uses 
https://github.com/tensorflow/tensorflow/blob/d79ba958f74c396af8ad1795b2bd3716d4434e21/tensorflow/compiler/xla/client/lib/arithmetic.cc#L175
to compute the argmin/argmax.

The way this function works is use `reduce` to find the min/max element and then use `EQ` to find the index of that element. However it seems like `reduce` actually change the value of the input slightly. If I added a 
```
return input - BroadcastInDim(reduced_input, input_shape.dimensions(), broadcast_dims);
```
in 
https://github.com/tensorflow/tensorflow/blob/d79ba958f74c396af8ad1795b2bd3716d4434e21/tensorflow/compiler/xla/client/lib/arithmetic.cc#L202

I would expect one of the return value being zero because `reduced_input` came from input but instead I saw

```
printing bf16[5,5] {
  { 0.0001802, 0.01196, 0.01361, 0.06934, 0.1021 },
  { -0.0003757, 0.2559, 3.147e-05, -0.0003929, -0.0004005 },
  { -0.0004005, 0.2559, -0.0003986, -0.0004005, -0.0004005 },
  { -0.0001755, -0.0001755, 0.2539, 0.0005188, -0.0001755 },
  { 0.09912, 0.00037, 0.0004959, 0.00037, 0.1074 }
}
```

If I do a `xm.mark_step` and then rerun the compution I saw
```
printing bf16[5,5] {
  { 0, 0.01172, 0.01367, 0.06934, 0.1016 },
  { 0, 0.2539, 0, 0, 0 },
  { 0, 0.2539, 0, 0, 0 },
  { 0, 0, 0.2539, 0.0009766, 0 },
  { 0.09961, 0, 0.0009766, 0, 0.1074 }
}
```

My best guess is that without `mark_step` xla is running in a mode that precision of bfloat16 doesn't get handle very well. I will check with XLA folks but this one should not affect the real use case.
